### PR TITLE
WIP:ORLP-439 REST API

### PR DESCRIPTION
**Pull request by Anblak**

### CODE REVIEW

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardImageController.java [hunk 1]
```diff
@@ -15,7 +15,7 @@
     private CardImageService cardImageService;
 
     @ResponseStatus(HttpStatus.OK)
-    @DeleteMapping(value = "/api/private/cardImage/{id}")
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/controller/CardImageController.java]: in other controllers we have `/` before `api`

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardRatingController.java [hunk 2]
```diff
@@ -32,35 +31,31 @@
...
      * @return CardRatingDTO
      */
-    @GetMapping("api/rate/card/{cardId}")
-    public ResponseEntity<CardRatingPublicDTO> getCardRatingById(@PathVariable Long cardId) {
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping
+    public CardRatingPublicDTO getCardRatingById(@PathVariable Long cardId) {
         CardRating cardRating = cardRatingService.getCardRatingById(cardId);
         Link selfLink = linkTo(methodOn(CardRatingController.class)
                 .getCardRatingById(cardRating.getId())).withRel("cardRating");
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/controller/CardRatingController.java]: We should use `withSelfRel`, if we want to build link to indicate current resource.

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardsFileController.java [hunk 3]
```diff
@@ -1,71 +1,46 @@
...
 
-    @Autowired
-    private CardLoadService cardLoadService;
-
-    @PostMapping("api/private/upload/deck/{deckId}/cards")
-    public ResponseEntity uploadFile(@RequestParam("file") MultipartFile cardsFile, @PathVariable Long deckId)
-            throws WrongFormatException,NotOwnerOperationException, NotAuthorisedUserException, EmptyFileException,
-            IOException {
+    @PostMapping("api/upload/deck/{deckId}/cards")
+    @ResponseStatus(HttpStatus.OK)
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/controller/CardsFileController.java]: Maybe better to return status `CREATED` or `NO_CONTENT`, but not sure)
Or return list of cards, that was uploaded)

File: src/main/java/com/softserve/academy/spaced/repetition/controller/dto/impl/DeckPrivateDTO.java [hunk 4]
```diff
@@ -11,7 +12,7 @@
 public class DeckPrivateDTO extends DTO<Deck> {
     public DeckPrivateDTO(Deck deck, Link link) {
         super(deck, link);
-        add(linkTo(methodOn(DeckController.class).getCardsByDeck(getEntity().getId())).withRel("cards"));
+        add(linkTo(methodOn(CardController.class).getCardsByDeck(getEntity().getId())).withRel("cards"));
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/controller/dto/impl/DeckPrivateDTO.java]: Not sure, but maybe better to build links inside of controller methods

File: src/main/java/com/softserve/academy/spaced/repetition/service/impl/CardServiceImpl.java [hunk 5]
```diff
@@ -82,29 +82,21 @@ public Card getCard(Long id) {
...
-                || card.getQuestion().trim().isEmpty()) {
-            throw new IllegalArgumentException(messageSource.getMessage("message.exception.cardFieldsNotEmpty",
-                    new Object[]{}, locale));
-        }
-        Deck deck = deckRepository.findOne(deckId);
-        card.setDeck(deck);
-        deck.getCards().add(cardRepository.save(card));
+        card.setDeck(deckRepository.findOne(deckId));
+        cardRepository.save(card);
         cardImageService.addCardImage(imageList, card);
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/service/impl/CardServiceImpl.java]: If we add some object, we should return this object (not sure)

File: src/main/java/com/softserve/academy/spaced/repetition/controller/dto/impl/DeckPrivateDTO.java (hunk 4 above)
Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/dto/impl/DeckPrivateDTO.java]: it`s link for another resources

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardImageController.java (hunk 1 above)
Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardImageController.java]: if / is not necessary on start of link, why we must put them?

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardRatingController.java (hunk 2 above)
Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardRatingController.java]: this controllers never used, so i did`t pay attention  for this controllers

File: src/main/java/com/softserve/academy/spaced/repetition/service/impl/CardServiceImpl.java (hunk 5 above)
Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/service/impl/CardServiceImpl.java]: it returned by controller in other way

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardsFileController.java (hunk 3 above)
Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardsFileController.java]: Must be status 2** , one oh them - OK

File: src/main/java/com/softserve/academy/spaced/repetition/controller/dto/impl/DeckPrivateDTO.java (hunk 4 above)
Comment (MoGUREV) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/dto/impl/DeckPrivateDTO.java]: I understand, i mean, that this link should be build inside of controller method.

File: src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java [hunk 6]
```diff
@@ -1,201 +1,100 @@
...
-    public ResponseEntity<CardPublicDTO> addCardByCategoryAndDeck(@PathVariable Long categoryId,
-                                                                  @PathVariable Long deckId,
-                                                                  @RequestBody CardDTO card) {
-        LOGGER.debug("Add card to categoryId: {}, deckId: {}", categoryId, deckId);
+    @PostMapping(value = "cards")
+    @ResponseStatus(HttpStatus.CREATED)
+    public CardPublicDTO addCard(@PathVariable Long deckId, @Validated @RequestBody CardDTO card) {
+        LOGGER.debug("Add card to deckId: {}", deckId);
         Card newCard = new Card(card.getTitle(), card.getQuestion(), card.getAnswer());
         cardService.addCard(newCard, deckId, card.getImages());
```
Comment (MoGUREV) [inline src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: Why didn't we get object of `Card` from method `addCard`?

Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: i just use created card (newCard), one of ways to write it

Comment (MoGUREV) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: We can get, for example, another object from method `addCard`, that not be equal to `newCard` object.

Comment (Anblak) [reply src/main/java/com/softserve/academy/spaced/repetition/controller/CardController.java]: we need just data of this object

**Pull request merged**